geometry-builder.cpp

```diff
@@ -760,8 +760,68 @@ void geometry_builder::set_exclude_broken_polygon(int exclude)
     excludepoly = exclude;
 }
 
+void geometry_builder::set_reprojection(struct reprojection *r)
+{
+    reprojection = r;
+}
+
+/**
+ * Computes area of given polygonal geometry.
+ * \return the area in projected units, or in EPSG 3857 if area reprojection is enabled
+ */
+double geometry_builder::getArea(const geos::geom::Geometry *geom) const
+{
+    // reprojection is not necessary, or has not been asked for.
+    if (!reprojection) {
+        return geom->getArea();
+    }
+
+    // MultiPolygon - return sum of individual areas
+    if (const auto *multi = dynamic_cast<const geos::geom::MultiPolygon *>(geom)) {
+        return std::accumulate(multi->begin(), multi->end(), 0.0,
+                               [=](double a, const geos::geom::Geometry *geom) { return a + getArea(geom); });
+    }
+
+    const auto *poly = dynamic_cast<const geos::geom::Polygon *>(geom);
+    if (!poly) {
+        return 0.0;
+    }
+
+    // standard polygon - reproject rings individually, then assemble polygon and
+    // compute area.
+
+    const auto *ext = poly->getExteriorRing();
+    auto *projectedExt = reproject_linearring(ext);
+    auto nholes = poly->getNumInteriorRing();
+    std::vector<geos::geom::Geometry *> *projectedHoles = new std::vector<geos::geom::Geometry *>(nholes);
```

Comment:
Wrap in std::unique_ptr<> and use release() below in createPolygon() to ensure memory gets clean up when an exception is thrown.